### PR TITLE
[extraction] Support extraction of Coq's string type to OCaml's string type

### LINKED ISSUE #2747: Strings extraction to OCaml native strings
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#2747
From: @gdsfh
Reported version: 8.3

### PR DESCRIPTION
**Pull request by xavierleroy**
**Kind:** feature

By default, Coq's strings extract to an OCaml type isomorphic to lists of 8-tuples of Booleans.  With the `OcamlExtrString` module required, Coq's characters (type `ascii`) extract to OCaml's native characters (instead of 8-tuples of Booleans), and Coq's strings (type `string`) extract to lists of characters.  This is better but still not as efficient as OCaml's strings.

This PR adds another extraction mode, triggered by requiring the new module  `OcamlExtrStringPlus` (module name open to discussion).  Then, Coq's strings extract to OCaml's strings.

In the new mode, some of the Coq functions from the String module are given efficient OCaml implementations over OCaml strings.  Other functions such as `String.get` are not, because they involve integers (type `nat`), whose extraction to OCaml varies depending on other extraction options.  Multiple OCaml implementations could be provided if it is deemed important.

Fixes #2747.

TODO if there is interest in this PR:
- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: plugins/extraction/ExtrOcamlStringPlus.v
Comment (SkySkimmer) [inline plugins/extraction/ExtrOcamlStringPlus.v]: Isn't this redundant with the one at line 55?

Comment (xavierleroy) [reply plugins/extraction/ExtrOcamlStringPlus.v]: Definitely!  Cut-and-paste error...

**Comment by gares** [comment]
At some point we started compiling plugins with warn-error...
```
OCAMLC    plugins/extraction/common.ml
File "plugins/extraction/common.ml", line 696, characters 8-25:
Warning 39: unused rec flag.
OCAMLC    plugins/cc/ccproof.mli
File "plugins/extraction/common.ml", line 1:
Error: Some fatal warnings were triggered (1 occurrences)
```

File: plugins/extraction/common.ml
Comment (ejgallego) [inline plugins/extraction/common.ml]: This `rec` is superfluous and making the build fail.

**Comment by JasonGross** [comment]
Very nice!

It might be nice to have some output tests for this.

Also, I'd suggest adding modules like `ExtrOcamlStringPlusNatInt` and `ExtrOcamlStringPlusNatBigInt` or whatever, and providing efficient implementations of the other functions, which `Export`s `ExtrOcamlStringPlus` and `ExtrOcamlNatInt` or whatever

**Comment by xavierleroy** [comment]
I fixed  the problems found by review and by CI testing.  I won't be able to work further on this PR in the coming weeks.

**Comment by pi8027** [comment]
This PR would be able to close #2747. Thank you!

**Comment by cpitclaudel** [comment]
Very nice job.  This will make extracted code much nicer to look at :)

**Comment by ppedrot** [comment]
Probably won't be ready in time, postponing.

**Comment by maximedenes** [comment]
Rebased.

File: plugins/extraction/common.ml
Comment (maximedenes) [inline plugins/extraction/common.ml]: I think we should use the `coqlib` machinery here, rather than hardcoding the path. I can take care of it.

**Review by maximedenes — COMMENTED** [review]
This looks great! Any reason to keep the old `ExtrOcamlString`? Is it just for compatibility? If we do want to keep it, maybe `ExtrOcamlNativeString` would be a more meaningful name for the new one.

**Comment by Zimmi48** [comment]
> maybe `ExtrOcamlNativeString` would be a more meaningful name for the new one

Or even better... `ExtrOCamlNativeString`

**Comment by maximedenes** [comment]
> Or even better... `ExtrOCamlNativeString`

It would be very confusing to have `ExtrOcamlString` and `ExtrOCamlNativeString`.

File: plugins/extraction/common.ml
Comment (maximedenes) [reply plugins/extraction/common.ml]: Done.

File: plugins/extraction/ExtrOcamlStringPlus.v
Comment (herbelin) [inline plugins/extraction/ExtrOcamlStringPlus.v]: I did not know `List.to_seq`. Browsing the web, I realize that it is new from 4.07, so this code requires OCaml 4.07, but, maybe Coq 8.12 shall also require it actually.

**Review by herbelin — APPROVED** [review]
Hi Xavier, I made a quick reading. That's great, and very awaited.

Let me then lay "à la cantonade" some ideas for possible extensions to whoever might be interested.

It seems to me that this could eventually be combined with the primitive string notations (implemented by @jasongross, #8965). At first view, a possibility would be to add a field `MLstring` to `ml_ast` and to hack `extract_term` so as to recognize Coq terms that denotes constant strings in some registered Coq type in which these constant strings live - not necessarily `Coq.Strings.String.string`. A generalization of `check_extract_string` could test this. The primitive string printer is `Notation.uninterp_prim_token` and it would do the same kind of job as  `get_native_string` and `is_native_string` do (`uninterp_prim_token` is working on `glob_constr`, so, first, a `detyping` would be needed). We would also need, similarly to what is done in `ExtractOcamlNativeString` for `Coq.Strings.String.string`,  to add specific `Extract` command for the specific high-level operations on the given type

To continue laying ideas for the future, it seems to me that this could also be extended to strings in patterns. I suspect that the generic function `Notation.uninterp_prim_token_cases_pattern` could do the work uniformly for all types equipped with a parser/printer for constant string patterns, to the price of adding a primitive string constructor in the extraction type `ml_pattern`.

I'm otherwise sharing the other comments in the discussion.

File: plugins/extraction/ExtrOcamlStringPlus.v
Comment (Zimmi48) [reply plugins/extraction/ExtrOcamlStringPlus.v]: Given the current discussions, I'd say it's quite likely that we'll continue requiring OCaml 4.05+ for Coq 8.12... So it would be better to keep this code compatible with OCaml 4.05.

File: plugins/extraction/common.ml
Comment (xavierleroy) [reply plugins/extraction/common.ml]: Fixed.

File: plugins/extraction/ExtrOcamlStringPlus.v
Comment (xavierleroy) [reply plugins/extraction/ExtrOcamlStringPlus.v]: Done.

File: plugins/extraction/ExtrOcamlStringPlus.v
Comment (xavierleroy) [reply plugins/extraction/ExtrOcamlStringPlus.v]: OK, I provided another implementation of the string <-> char list conversions that should work in 4.05.

**Comment by xavierleroy** [comment]
Thanks to everyone who reviewed or pushed updates.  The latest version of this PR uses `ExtrOcamlNativeString.v` instead of `ExtrOcamlStringPlus.v` (much better, thanks), and addresses the latest review comments.

**Comment by mattam82** [comment]
It seems the code now requires String and Ascii to be loaded for any extraction as it expects the `Register` effects to be performed before calling extraction.

**Comment by maximedenes** [comment]
I added a few commits:
- improved extraction of string literal in Haskell (just did the same as for OCaml)
- factorized the common parts of `ExtrOcamlString` and `ExtrOcamlNativeString`
- added documentation
- added a test case
- added a changelog entry

@xavierleroy don't hesitate to tweak the changelog entry, if you wish. Let me know if you don't like some of the other changes.

**Comment by maximedenes** [comment]
@herbelin This should be ready, modulo CI and a go from @xavierleroy. Can you be the assignee, since I ended up committing some code in the PR?

**Comment by pi8027** [comment]
@maximedenes @herbelin Since I recently became a secondary maintainer of the extraction plugin, I can be the assignee.

File: plugins/extraction/ExtrOcamlNativeString.v
Comment (pi8027) [inline plugins/extraction/ExtrOcamlNativeString.v]: Same here.

**Review by pi8027 — COMMENTED** [review]
Almost LGTM, but there are a few remarks. Extraction to Haskell's string type/literal probably should also support `ByteString` and/or `[Word8]` (#7870), but this can be done in another PR.

File: plugins/extraction/ExtrOcamlNativeString.v
Comment (xavierleroy) [reply plugins/extraction/ExtrOcamlNativeString.v]: No, because `List.init` was introduced in OCaml 4.06 and I was given strict orders to be compatible with OCaml 4.05.

File: plugins/extraction/ExtrOcamlNativeString.v
Comment (xavierleroy) [reply plugins/extraction/ExtrOcamlNativeString.v]: Same reason.

File: plugins/extraction/ExtrOcamlChar.v
Comment (maximedenes) [reply plugins/extraction/ExtrOcamlChar.v]: Ok, but this code was not changed in this PR, just moved, so let's discuss this somewhere else.

File: doc/sphinx/addendum/extraction.rst
Comment (xavierleroy) [inline doc/sphinx/addendum/extraction.rst]: Is this change of name (string -> string__1) really intended? or some automatic renaming gone wild?

Comment (maximedenes) [reply doc/sphinx/addendum/extraction.rst]: I did it so that I could refer to that particular string in the description that comes just after. It is used in some other places of the manual (the `1` will appear as a subscript).

File: plugins/extraction/ExtrOcamlNativeString.v
Comment (pi8027) [reply plugins/extraction/ExtrOcamlNativeString.v]: Sure.

File: plugins/extraction/ExtrOcamlChar.v
Comment (pi8027) [reply plugins/extraction/ExtrOcamlChar.v]: OK. I will open an issue or PR.

**Review by pi8027 — APPROVED** [review]
LGTM. I will merge as soon as @xavierleroy gives a go sign.

**Comment by xavierleroy** [comment]
Sorry, I didn't realize you were waiting on my green light.   The PR is fine with me and I'll be happy to see it merged.

**Comment by pi8027** [comment]
@xavierleroy Never mind. Thank you again for the great improvement!

**Pull request merged**
